### PR TITLE
[WIP]migrator: handle function hoist API changes. rdar://31234806

### PR DESCRIPTION
**Pull request by nkcsgexi**

### CODE REVIEW

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

File: include/swift/IDE/APIDigesterData.h [hunk 1]
```diff
@@ -198,28 +199,51 @@ struct CommonDiffItem: public APIDiffItem {
...
   StringRef usr;
   StringRef newTypeName;
   StringRef newPrintedName;
   Optional<uint8_t> selfIndex;
+  Optional<uint8_t> removedIndex;
   StringRef oldPrintedName;
 
+private:
+  DeclNameViewer OldNameViwer;
+  DeclNameViewer NewNameViwer;
```
Comment (nathawes) [inline include/swift/IDE/APIDigesterData.h]: Typo: Viwer

File: lib/Migrator/APIDiffMigratorPass.cpp [hunk 2]
```diff
@@ -327,17 +352,76 @@ struct APIDiffMigratorPass : public ASTMigratorPass, public SourceEntityWalker {
...
+      getCallArgInfo(SM, Arg, LabelRangeEndAt::LabelNameOnly);
+    if (!AllArgs.size())
+      return false;
+    DeclNameViewer NewName = Item->getNewName();
+    llvm::SmallVector<unsigned, 2> IgnoredArgIndices;
+    IgnoredArgIndices.push_back(*Item->selfIndex);
+    if (auto RI = Item->removedIndex)
+      IgnoredArgIndices.push_back(*RI);
+    emitRenameLabelChanges(Arg, NewName, IgnoredArgIndices);
+    auto *SelfExpr = AllArgs[0].ArgExp;
```
Comment (nathawes) [inline lib/Migrator/APIDiffMigratorPass.cpp]: Does this assume SelfIndex is 0? Would an assert be worthwhile?

Comment (nkcsgexi) [inline lib/Migrator/APIDiffMigratorPass.cpp]: good catch, line 370 will return if SelfIndex is not 0. Current transformation mechanism may have limitation to handle non-zero selfindex. Fortunately, we don't see any of them in real frameworks.

**Review by nathawes — APPROVED** [review]
LGTM!

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Pull request merged**